graphql_compiler/query_formatting/common.py

```diff
@@ -74,49 +56,55 @@ def _deserialize_anonymous_json_argument(expected_type: GraphQLScalarType, value
-            )
-
-    # Check if the type of the value is correct
-    correct_type = True
-    expected_python_types = allowed_types_for_graphql_type[expected_type.name]
-    if isinstance(value, bool) and not is_same_type(GraphQLBoolean, expected_type):
-        correct_type = False  # We explicitly disallow passing boolean values for non-boolean types
-    if not isinstance(value, expected_python_types):
-        correct_type = False
-    if not correct_type:
-        raise ValueError(
-            "Unexpected type {}. Expected one of {}.".format(type(value), expected_python_types)
+def _validate_json_scalar_argument(name: str, expected_type: GraphQLScalarType, value: Any) -> None:
+    """Validate that the JSON serialized scalar argument has the expected python type."""
+    expected_python_types = _ALLOWED_JSON_SCALAR_TYPES.get(expected_type.name)
+    if expected_python_types is None:
+        raise AssertionError(
+            f"Got unsupported GraphQL type {expected_type} for argument {name} with value {value}."
         )
+    elif (
+        # We explicitly disallow passing boolean values for non-boolean types
+        (isinstance(value, bool) and not is_same_type(GraphQLBoolean, expected_type))
+        or not isinstance(value, expected_python_types)
+    ):
+        _raise_invalid_type_error(name, expected_python_types, value)
+
+
+_CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS = MappingProxyType(
+    {GraphQLInt.name: int, GraphQLFloat.name: float}
+)
+_SCALAR_DESERIALIZATION_FUNCTIONS = MappingProxyType(
+    {
+        name: _CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS.get(name, graphql_type.parse_value)
+        for name, graphql_type in SCALAR_TYPE_NAME_TO_VALUE.items()
+    }
+)
 
-    name_to_custom_type = {graphql_type.name: graphql_type for graphql_type in CUSTOM_SCALAR_TYPES}
-    # Parse the value. In most cases we can use the default GraphQL parser, but there are some
-    # special cases where we are more permissive than the default.
-    # By default, strings cannot be parsed to float.
-    if is_same_type(expected_type, GraphQLFloat):
-        return float(value)
-    # Use the default GraphQL parser to parse the value
-    elif expected_type.name in name_to_custom_type:
-        # Since we cannot serialize the parse_value function of custom scalar types when
-        # serializing a schema, it is possible that the parse_value is incorrectly set. We must,
-        # therefore, use the parse_value function of the original scalar type definition.
-        return name_to_custom_type[expected_type.name].parse_value(value)
-    else:
-        return expected_type.parse_value(value)
+
+def _deserialize_json_scalar_argument(name, expected_type: GraphQLScalarType, value: Any) -> Any:
+    """Deserialize the json serialized scalar argument."""
+    _validate_json_scalar_argument(name, expected_type, value)
+    try:
+        return _SCALAR_DESERIALIZATION_FUNCTIONS[expected_type.name](value)
+    except (ValueError, TypeError) as e:
```

Comment:
How about `KeyError` because of `_SCALAR_DESERIALIZATION_FUNCTIONS[expected_type.name]`? Is that a bug in the code, or a user input error?

You could avoid this by applying a variant of the "parse, don't validate" principle: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

You have chosen to call `_validate_json_scalar_argument()` which returns no arguments, and then use that as an implicit "proof" that `_SCALAR_DESERIALIZATION_FUNCTIONS[expected_type.name]` won't raise `KeyError`. In addition to being implicit and therefore worse than explicit, this code is also not particularly stable in the face of refactoring and code evolution.

Instead, make a new function instead of `_validate_json_scalar_argument()` that has to either raise an error or return a deserializer function, which you can then call inside the `try-except`. That way, this new function has to do the potentially-KeyError-producing step, but in that case KeyError is an obvious bug due to a failure to appropriately check the user's input.

Reply:
Thanks for the article link. It was really good. Also thanks for the suggestion, I think that the code is much cleaner now.